pkg/registry/common/connect/nse_server.go

```diff
@@ -0,0 +1,132 @@
+}
+
+type connectNSEServer struct {
+	dialOptions       []grpc.DialOption
+	clientFactory     func(ctx context.Context, cc grpc.ClientConnInterface) registry.NetworkServiceEndpointRegistryClient
+	cache             map[string]*nseCacheEntry
+	cacheLock         sync.Mutex
+	ctx               context.Context
+	once              sync.Once
+	connectExpiration time.Duration
+}
+
+// NewNetworkServiceEndpointRegistryServer creates new connect NetworkServiceEndpointRegistryServer with specific chain context, registry client factory and options
+// that allows connecting to other registries via passed clienturl.
+func NewNetworkServiceEndpointRegistryServer(
+	chainContext context.Context,
+	clientFactory func(ctx context.Context, cc grpc.ClientConnInterface) registry.NetworkServiceEndpointRegistryClient,
+	options ...Option) registry.NetworkServiceEndpointRegistryServer {
+	r := &connectNSEServer{
+		clientFactory:     clientFactory,
+		cache:             map[string]*nseCacheEntry{},
+		ctx:               chainContext,
+		connectExpiration: time.Minute,
+	}
+	for _, o := range options {
+		o.apply(r)
+	}
+	return r
+}
+
+func (c *connectNSEServer) Register(ctx context.Context, ns *registry.NetworkServiceEndpoint) (*registry.NetworkServiceEndpoint, error) {
+	c.once.Do(c.init)
+	return c.connect(ctx).Register(ctx, ns)
+}
+
+func (c *connectNSEServer) Find(q *registry.NetworkServiceEndpointQuery, s registry.NetworkServiceEndpointRegistry_FindServer) error {
+	c.once.Do(c.init)
+	return adapters.NetworkServiceEndpointClientToServer(c.connect(s.Context())).Find(q, s)
+}
+
+func (c *connectNSEServer) Unregister(ctx context.Context, ns *registry.NetworkServiceEndpoint) (*empty.Empty, error) {
+	c.once.Do(c.init)
+	return c.connect(ctx).Unregister(ctx, ns)
+}
+
+func (c *connectNSEServer) init() {
+	go func() {
+		defer func() {
+			c.cacheLock.Lock()
+			c.cache = nil
+			c.cacheLock.Unlock()
+		}()
+		for {
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-time.After(c.connectExpiration):
+				c.cacheLock.Lock()
+				var deleteList []string
+				for k, v := range c.cache {
+					if time.Until(v.expiryTime) <= 0 {
+						deleteList = append(deleteList, k)
+					}
+				}
+				for _, item := range deleteList {
+					delete(c.cache, item)
+				}
+				c.cacheLock.Unlock()
+			}
+		}
+	}()
+}
+
+func (c *connectNSEServer) connect(ctx context.Context) registry.NetworkServiceEndpointRegistryClient {
+	key := ""
+	if url := clienturl.ClientURL(ctx); url != nil {
+		key = url.String()
+	}
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	if v, ok := c.cache[key]; ok && v != nil {
+		v.expiryTime = time.Now().Add(c.connectExpiration)
+		return v.client
+	}
+	r := clienturl.NewNetworkServiceEndpointRegistryClient(ctx, c.clientFactory, c.dialOptions...)
+	c.cache[key] = &nseCacheEntry{client: r, expiryTime: time.Now().Add(c.connectExpiration)}
+	return r
+}
+
+func (c *connectNSEServer) setExpirationDuration(d time.Duration) {
```

Comment:
Why not just set these directly inline rather than making new functions that simply obscure what's being done?

Reply:
It is needed to allow use the same options for ns/nse servers. In the case of inline, we'll need to create duplicated options and it will look something like that
```
NewNetworkServiceRegistryServer(..., connect.WithNetworkServiceClientDialOptions(..))
```
So to prevent this complication, I've added these functions.

Reply:
OK... is there an event driven rather than timer driven option we could use to decide whether or not to retain connections?

Reply:
I like idea of simply store connection during some period while it is using.

Reply:
So here's the question... what triggers needing a client and what triggers no longer needing it?

For example... in NetworkService connect, that's driven by whether or not we have active NetworkService.Connections to a particular client.   In other words, do we have active interests with that client.  What is the 'active' interest here?

Reply:
> It is needed to allow use the same options for ns/nse servers. In the case of inline, we'll need to create duplicated options and it will look something like that
> 
> ```
> NewNetworkServiceRegistryServer(..., connect.WithNetworkServiceClientDialOptions(..))
> NewNetworkServiceEndpointRegistryServer(..., connect.WithNetworkServiceEndpointClientDialOptions(..))
> ```
> 
> So to prevent this complication, I've added these functions.

Also... I don't see why Options requires having these tiny one line functions that obscure what's happening attached to the connectNSEServer... perhaps I'm missing something.

Reply:
> What is the 'active' interest here?

I think it is `active` while the client is interested to use it.

Reply:
>Also... I don't see why Options requires having these tiny one line functions that obscure what's happening attached to the connectNSEServer... perhaps I'm missing something.

Hm... Probably it is correct we can just pass dial options and expiration duration via function parameters :)

Reply:
As a mechanical matter, that would be a better approach, definitely, but the question fundamentally remains: what constitutes functionally the meaningful lifetime of needing that connection?  My point is, I am pretty sure that's *event* not *time* driven.

Reply:
> My point is, I am pretty sure that's event not time driven.

But here is many things that can be not controlled for example in time _t1_ result of DNS query "domain1" can be url1 in time _t2_ result of DNS query "domain1" can be url2 and so on.

Reply:
Finally, options are removed and it looks more clear, thanks :)